File: typescript/src/gyro/gyroPoolMath.ts
import { MathSol, WAD } from '../utils/math';

export class GyroPoolMath {
    static _SQRT_1E_NEG_1 = 316227766016837933n;
    static _SQRT_1E_NEG_3 = 31622776601683793n;
    static _SQRT_1E_NEG_5 = 3162277660168379n;
    static _SQRT_1E_NEG_7 = 316227766016837n;
    static _SQRT_1E_NEG_9 = 31622776601683n;
    static _SQRT_1E_NEG_11 = 3162277660168n;
    static _SQRT_1E_NEG_13 = 316227766016n;
    static _SQRT_1E_NEG_15 = 31622776601n;
    static _SQRT_1E_NEG_17 = 3162277660n;

    /// @dev Implements a square root algorithm using Newton's method and a first-guess optimization.
    static sqrt(input: bigint, tolerance: bigint): bigint {
        if (input === 0n) {
            return 0n;
        }

        let guess = this._makeInitialGuess(input);

        // At this point `guess` is an estimation with one bit of precision. We know the true value is a uint128,
        // since it is the square root of a uint256. Newton's method converges quadratically (precision doubles at
        // every iteration). We thus need at most 7 iterations to turn our partial result with one bit of precision
        // into the expected uint128 result.
        guess = (guess + (input * WAD) / guess) / 2n;
        guess = (guess + (input * WAD) / guess) / 2n;
        guess = (guess + (input * WAD) / guess) / 2n;
        guess = (guess + (input * WAD) / guess) / 2n;
        guess = (guess + (input * WAD) / guess) / 2n;
        guess = (guess + (input * WAD) / guess) / 2n;
        guess = (guess + (input * WAD) / guess) / 2n;

        // Check that squaredGuess (guess * guess) is close enough from input. `guess` has less than 1 wei error, but
        // the loss of precision in the 18 decimal representation causes an error in the squared number, which must be
        // less than `(guess * tolerance) / WAD`. Tolerance, in this case, is a very small number (< 10),
        // so the tolerance will be very small too.
        const guessSquared = MathSol.mulDownFixed(guess, guess);
        if (
            !(
                guessSquared <= input + MathSol.mulUpFixed(guess, tolerance) &&
                guessSquared >= input - MathSol.mulUpFixed(guess, tolerance)
            )
        ) {
            throw Error('_sqrt FAILED');
        }

        return guess;
    }

    static _makeInitialGuess(input: bigint): bigint {
        if (input >= WAD) {
            return (1n << this._intLog2Halved(input / WAD)) * WAD;
        } else {
            if (input <= 10n) return this._SQRT_1E_NEG_17;
            if (input <= 100n) return 10n ** 10n;
            if (input <= 1000n) return this._SQRT_1E_NEG_15;
            if (input <= 10000n) return 10n ** 11n;
            if (input <= 100000n) return this._SQRT_1E_NEG_13;
            if (input <= 1000000n) return 10n ** 12n;
            if (input <= 10000000n) return this._SQRT_1E_NEG_11;
            if (input <= 100000000n) return 10n ** 13n;
            if (input <= 1000000000n) return this._SQRT_1E_NEG_9;
            if (input <= 10000000000n) return 10n ** 14n;
            if (input <= 100000000000n) return this._SQRT_1E_NEG_7;
            if (input <= 1000000000000n) return 10n ** 15n;
            if (input <= 10000000000000n) return this._SQRT_1E_NEG_5;
            if (input <= 100000000000000n) return 10n ** 16n;
            if (input <= 1000000000000000n) return this._SQRT_1E_NEG_3;
            if (input <= 10000000000000000n) return 10n ** 17n;
            if (input <= 100000000000000000n) return this._SQRT_1E_NEG_1;
            return input;
        }
    }

    static _intLog2Halved(x: bigint): bigint {
        let n = 0n; // Initialize n as a BigInt

        if (x >= 1n << 128n) {
            x >>= 128n;
            n += 64n;
        }
        if (x >= 1n << 64n) {
            x >>= 64n;
            n += 32n;
        }
        if (x >= 1n << 32n) {
            x >>= 32n;
            n += 16n;
        }
        if (x >= 1n << 16n) {
            x >>= 16n;
            n += 8n;
        }
        if (x >= 1n << 8n) {
            x >>= 8n;
            n += 4n;
        }
        if (x >= 1n << 4n) {
            x >>= 4n;
            n += 2n;
        }
        if (x >= 1n << 2n) {
            x >>= 2n;
            n += 1n;
        }

        return n;
    }
}

File: rust/src/pools/gyro/gyro_pool_math.rs
use crate::common::constants::WAD;
use crate::common::maths::{mul_down_fixed, mul_up_fixed};
use alloy_primitives::{uint, U256};

// Square root constants for different precision levels
pub const SQRT_1E_NEG_1: U256 = uint!(316227766016837933_U256);
pub const SQRT_1E_NEG_3: U256 = uint!(31622776601683793_U256);
pub const SQRT_1E_NEG_5: U256 = uint!(3162277660168379_U256);
pub const SQRT_1E_NEG_7: U256 = uint!(316227766016837_U256);
pub const SQRT_1E_NEG_9: U256 = uint!(31622776601683_U256);
pub const SQRT_1E_NEG_11: U256 = uint!(3162277660168_U256);
pub const SQRT_1E_NEG_13: U256 = uint!(316227766016_U256);
pub const SQRT_1E_NEG_15: U256 = uint!(31622776601_U256);
pub const SQRT_1E_NEG_17: U256 = uint!(3162277660_U256);

/// Implements a square root algorithm using Newton's method and a first-guess optimization.
/// Based on the Python implementation in gyro_pool_math.py
pub fn gyro_pool_math_sqrt(x: &U256, tolerance: u64) -> U256 {
    if x.is_zero() {
        return U256::ZERO;
    }

    let mut guess = make_initial_guess(x);

    // Perform Newton's method iterations
    for _ in 0..7 {
        guess = (guess + (x * WAD) / guess) / U256::from(2u64);
    }

    // Check that squaredGuess (guess * guess) is close enough from input
    let guess_squared =
        mul_down_fixed(&guess, &guess).unwrap_or_else(|_| panic!("mul_down_fixed failed"));
    let tolerance_bigint = U256::from(tolerance);

    let upper_bound = x + mul_up_fixed(&guess, &tolerance_bigint)
        .unwrap_or_else(|_| panic!("mul_up_fixed failed"));
    let lower_bound = x - mul_up_fixed(&guess, &tolerance_bigint)
        .unwrap_or_else(|_| panic!("mul_up_fixed failed"));

    if !(guess_squared <= upper_bound && guess_squared >= lower_bound) {
        panic!("_sqrt FAILED");
    }

    guess
}

/// Makes an initial guess for the square root calculation
fn make_initial_guess(x: &U256) -> U256 {
    if x >= &WAD {
        let x_div_wad = x / WAD;
        let log2_halved = int_log2_halved(&x_div_wad);
        (U256::ONE << log2_halved) * WAD
    } else if x <= &U256::from(10u64) {
        SQRT_1E_NEG_17
    } else if x <= &U256::from(100u64) {
        U256::from(10_000_000_000u64)
    } else if x <= &U256::from(1000u64) {
        SQRT_1E_NEG_15
    } else if x <= &U256::from(10000u64) {
        U256::from(100_000_000_000u64)
    } else if x <= &U256::from(100000u64) {
        SQRT_1E_NEG_13
    } else if x <= &U256::from(1000000u64) {
        U256::from(1_000_000_000_000u64)
    } else if x <= &U256::from(10000000u64) {
        SQRT_1E_NEG_11
    } else if x <= &U256::from(100000000u64) {
        U256::from(10_000_000_000_000u64)
    } else if x <= &U256::from(1000000000u64) {
        SQRT_1E_NEG_9
    } else if x <= &U256::from(10000000000u64) {
        U256::from(100_000_000_000_000u64)
    } else if x <= &U256::from(100000000000u64) {
        SQRT_1E_NEG_7
    } else if x <= &U256::from(1000000000000u64) {
        U256::from(1_000_000_000_000_000u64)
    } else if x <= &U256::from(10000000000000u64) {
        SQRT_1E_NEG_5
    } else if x <= &U256::from(100000000000000u64) {
        U256::from(10_000_000_000_000_000u64)
    } else if x <= &U256::from(1000000000000000u64) {
        SQRT_1E_NEG_3
    } else if x <= &U256::from(10000000000000000u64) {
        U256::from(100_000_000_000_000_000u64)
    } else if x <= &U256::from(100000000000000000u64) {
        SQRT_1E_NEG_1
    } else {
        *x
    }
}

/// Calculates the integer logarithm base 2 divided by 2
fn int_log2_halved(x: &U256) -> u64 {
    let mut n = 0u64;
    let mut local_x = *x;

    if local_x >= U256::ONE << 128 {
        local_x >>= 128;
        n += 64;
    }
    if local_x >= U256::ONE << 64 {
        local_x >>= 64;
        n += 32;
    }
    if local_x >= U256::ONE << 32 {
        local_x >>= 32;
        n += 16;
    }
    if local_x >= U256::ONE << 16 {
        local_x >>= 16;
        n += 8;
    }
    if local_x >= U256::ONE << 8 {
        local_x >>= 8;
        n += 4;
    }
    if local_x >= U256::ONE << 4 {
        local_x >>= 4;
        n += 2;
    }
    if local_x >= U256::ONE << 2 {
        local_x >>= 2;
        n += 1;
    }

    n
}
